apycula/chipdb.py

```diff
@@ -1898,9 +1975,756 @@ def dat_portmap(dat, dev, device):
+                        if wire_idx < 0:
+                            continue
+                        wire = wirenames[wire_idx]
+                        nam = f'ACCLOAD{i}'
+                        # for aux cells create Himbaechel nodes
+                        wire_type = 'DSP_I'
+                        if wire.startswith('CLK') or wire.startswith('CE') or wire.startswith('LSR'):
+                            wire_type = 'TILE_CLK'
+                        create_port_wire(dev, row, col, off, bel, name, nam, wire, wire_type)
+
+                    # dat.portmap['AluIn'] and dat.portmap['AluInDlt'] indicate port offset in cells
+                    for i in range(len(dat.portmap['AluIn'])):
+                        off = dat.portmap['AluInDlt'][i][column]
+                        wire_idx = dat.portmap['AluIn'][i][column]
+                        if wire_idx < 0:
+                            continue
+                        wire = wirenames[wire_idx]
+                        # input wire sequence: A0-53, B0-53, CASI0-54
+                        # ASIGN, BSIGN
+                        if i in range(54):
+                            nam = f'A{i}'
+                        elif i in range(54, 108):
+                            nam = f'B{i - 54}'
+                        elif i in range(163, 165):
+                            nam = ['ASIGN', 'BSIGN'][i - 163]
+                        else:
+                            continue
+                        create_port_wire(dev, row, col, off, bel, name, nam, wire, "DSP_I")
+
+                    # outputs
+                    for i in range(len(dat.portmap['AluOut'])):
+                        off = dat.portmap['AluOutDlt'][i][column]
+                        wire_idx = dat.portmap['AluOut'][i][column]
+                        if wire_idx < 0:
+                            continue
+                        wire = wirenames[wire_idx]
+                        # output wire sequence:
+                        # DOUT0-54
+                        # unknown -1
+                        if i > 53:
+                            break
+                        nam = f'DOUT{i}'
+                        create_port_wire(dev, row, col, off, bel, name, nam, wire, "DSP_O")
+
+                elif name.startswith('MULT36X36'):
+                    # 36x36 are assembled from 4 18x18 multipliers and two
+                    # ALU54D.
+                    # macro:          0       |       1
+                    # mult18x18:   0     1    |   0       1
+                    # A:  0-17    0-17        |  0-17
+                    #    18-35          0-17  |          0-17
+                    # -----------------------------------------------------
+                    # B:  0-17    0-17  0-17  |
+                    #    18-35                |  0-17    0-17
+                    # The ALU54D outputs turned out to be the easiest to find.
+                    # outputs
```

Comment:
Would be cool to add some docs about their layout and connections at some point.

Reply:
Agree. Only a few of the diagrams in the DSP documentation shed any light on the implementation, and the diagram they provide for MULT36X36 is not one of them.
I need a good diagramming program.